contracts/Turnstone.vy

```diff
@@ -0,0 +1,108 @@
+    return signer == ecrecover(message_digest, sig.v, sig.r, sig.s)
+
+@internal
+def check_validator_signatures(current_valset: ValsetArgs, sigs: DynArray[Signature, MAX_VALIDATORS], hash: bytes32):
+    i: uint256 = 0
+    cumulative_power: uint256 = 0
+    for sig in sigs:
+        if sig.v != 0:
+            assert self.verify_signature(current_valset.validators[i], hash, sig), "Invalid Signature"
+            cumulative_power += current_valset.powers[i]
+            if cumulative_power >= POWER_THRESHOLD:
+                break
+        i += 1
+    assert cumulative_power >= POWER_THRESHOLD, "Insufficient Power"
+
+@internal
+@view
+def make_checkpoint(valset_args: ValsetArgs) -> bytes32:
+    return keccak256(_abi_encode(valset_args.validators, valset_args.powers, valset_args.nonce, TURNSTONE_ID, method_id=method_id("checkpoint(address[],uint256[],uint256,uint256)")))
+
+@external
+def update_valset(new_valset: ValsetArgs, current_valset: ValsetArgs, sigs: DynArray[Signature, MAX_VALIDATORS]):
+    assert new_valset.nonce > current_valset.nonce, "Invalid Valset Nonce"
+    cumulative_power: uint256 = 0
+    i: uint256 = 0
+    for validator in new_valset.validators:
+        cumulative_power += new_valset.powers[i]
+        if cumulative_power >= POWER_THRESHOLD:
+            break
+        i += 1
+    assert cumulative_power >= POWER_THRESHOLD, "Insufficient Power"
+    assert self.last_checkpoint == self.make_checkpoint(current_valset), "Incorrect Checkpoint"
+    new_checkpoint: bytes32 = self.make_checkpoint(new_valset)
+    self.check_validator_signatures(current_valset, sigs, new_checkpoint)
+    self.last_checkpoint = new_checkpoint
+
+@external
+def submit_logic_call(current_valset: ValsetArgs, sigs: DynArray[Signature, MAX_VALIDATORS], args: LogicCallArgs, message_id: uint256, deadline: uint256):
+    assert block.timestamp <= deadline, "Timeout"
+    assert self.last_checkpoint == self.make_checkpoint(current_valset), "Incorrect Checkpoint"
+    args_hash: bytes32 = keccak256(_abi_encode(args, message_id, deadline, method_id=method_id("logic_call((address,bytes),uint256,uint256)")))
+    self.check_validator_signatures(current_valset, sigs, args_hash)
+    raw_call(args.logicContractAddress, args.payload)
+
+@external
+def submit_batch_call(current_valset: ValsetArgs, sigs: DynArray[Signature, MAX_VALIDATORS], args: DynArray[LogicCallArgs, MAX_BATCH], message_id: uint256, deadline: uint256):
+    assert block.timestamp <= deadline, "Timeout"
+    assert self.last_checkpoint == self.make_checkpoint(current_valset), "Incorrect Checkpoint"
+    args_hash: bytes32 = keccak256(_abi_encode(args, message_id, deadline, method_id=method_id("batch_logic_call((address,bytes)[],uint256,uint256)")))
+    self.check_validator_signatures(current_valset, sigs, args_hash)
+    for arg in args:
+        raw_call(arg.logicContractAddress, arg.payload)
```

Comment:
what happens if the contract we are calling throws an exception (if that is even possible) or returns an error? Is our piece of code continue executing?
Also, is our contract that pays for gas that the underlying contract spent?

Reply:
- if the transaction we are calling from our contract fails, the entire transaction will fail. And the gas until the revert point should be paid. Fortunately, we can foresee if the transaction will fail or not before sending the transaction (using gas estimate)
- The contract can't pay the gas fee. Only a wallet or an account can pay the gas fee. That means NO transaction can occur from the contract.